### ISSUE TITLE
Predictions are too extreme and nonlinear

### ISSUE THREAD

**Issue by MaksimBocharenko** [microsoft/LightGBM#2247]
Hello,

I am using lightgbm regression model with mix of continuous and categorical variables.
The model build is dockerized and ran on AWS Fargate every day, saved as pickle in s3 and later used for predictions. The objective functions I used are "mape" or "quantile".
In very rare cases the predicted values are extremely out of range although input vectors are almost identical.

For example for the following two vectors (the minor difference in first value only) I receive predictions 1549.8055556033898 and 3.1485288540756817e+20 accordingly and I cannot find the reason for this problem. I was expecting for regression to be more smooth.

[1354.8666398525238, 166.6166398525238, 151.8666398525238, 6549.0,
 1176.0, 5939.683397849, 237.427891998, 6562.0, 2.0, 3.0, 1.0,
 2.0, 6549.0, 1176.0, 1.325975024, 152, 1, 0, 'a', 'b', 'c', 'd', 0, 1]

[1357.8666398525238, 166.6166398525238, 151.8666398525238, 6549.0,
 1176.0, 5939.683397849, 237.427891998, 6562.0, 2.0, 3.0, 1.0,
 2.0, 6549.0, 1176.0, 1.325975024, 152, 1, 0, 'a', 'b', 'c', 'd', 0, 1]

I cannot send data I have but I tried to generate locally data from similar distribution and size and received unexpected predictions as well, you can see predictions range much wider than actual target variable distribution. 

What I am missing or maybe doing something wrong? 

## Environment info

Operating System: macOS Mojave 10.14.4
CPU model: 3.5 GHz Intel Core i7
Python: 2.7.15
LightGBM: 2.2.2

## Reproducible examples

import lightgbm
import pandas as pd
import numpy as np

n = 500000
df = pd.DataFrame()
np.random.seed(0)
df['feat_1'] = np.random.normal(1500, 500, n)
df['feat_2'] = np.random.normal(500, 250, n)
df['feat_3'] = np.random.normal(500, 200, n)
df['feat_4'] = np.random.normal(6000, 3500, n)
df['feat_5'] = np.random.normal(1500, 200000, n)
df['feat_6'] = np.random.normal(1500, 50000, n)
df['feat_7'] = np.random.normal(6000, 3500, n)
df['feat_9'] = np.random.normal(6000, 3500, n)
df['feat_9'] = np.random.randint(1, 4, n)
df['feat_10'] = np.random.randint(2, 5, n)
df['feat_11'] = np.random.randint(1, 4, n)
df['feat_12'] = np.random.randint(0, 4, n)
df['feat_13'] = np.random.normal(6000, 3500, n)
df['feat_14'] = np.random.normal(1000, 500, n)
df['feat_15'] = np.random.normal(1.5, 0.5, n)
df['feat_16'] = np.random.randint(0, 200, n)
df['feat_17'] = np.random.randint(0, 6, n)
df['feat_18'] = np.random.randint(0, 23, n)
df['feat_19'] = pd.Categorical(np.random.choice(['a', 'b', 'c', 'd', 'e'], n), categories=['a', 'b', 'c', 'd', 'e'])
df['feat_20'] = pd.Categorical(np.random.choice(['a', 'b', 'c', 'd', 'e'], n), categories=['a', 'b', 'c', 'd', 'e'])
df['feat_21'] = pd.Categorical(np.random.choice(['a', 'b', 'c', 'd', 'e'], n), categories=['a', 'b', 'c', 'd', 'e'])
df['feat_22'] = pd.Categorical(np.random.choice(['a', 'b', 'c', 'd', 'e'], n), categories=['a', 'b', 'c', 'd', 'e'])
df['feat_23'] = pd.Categorical(np.random.choice([0, 1], n), categories=[0, 1])
df['feat_24'] = pd.Categorical(np.random.choice([0, 1], n), categories=[0, 1])
y = np.random.normal(1500, 500, n)

print('Y range', min(y), max(y))

model = lightgbm.LGBMRegressor(random_state=0, n_estimators=200, objective='mape')
fit = model.fit(df, y, verbose=True)
prediction = fit.predict(df)

print('Y prediction range', min(prediction), max(prediction))

Output:

('Y range', -778.1497475199135, 3781.263647592668)
('Y prediction range', -2.2727664646000095e+23, 2.5311693808179446e+22)

**Comment by guolinke** [comment]
I just run your code, in the latest master branch:
```
Python 3.7.0 (default, Jun 28 2018, 08:04:48) [MSC v.1912 64 bit (AMD64)] :: Anaconda, Inc. on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import lightgbm
>>> import pandas as pd
>>> import numpy as np
>>>
>>> n = 500000
>>> df = pd.DataFrame()
>>> np.random.seed(0)
>>> df['feat_1'] = np.random.normal(1500, 500, n)
>>> df['feat_2'] = np.random.normal(500, 250, n)
>>> df['feat_3'] = np.random.normal(500, 200, n)
>>> df['feat_4'] = np.random.normal(6000, 3500, n)
>>> df['feat_5'] = np.random.normal(1500, 200000, n)
>>> df['feat_6'] = np.random.normal(1500, 50000, n)
>>> df['feat_7'] = np.random.normal(6000, 3500, n)
>>> df['feat_9'] = np.random.normal(6000, 3500, n)
>>> df['feat_9'] = np.random.randint(1, 4, n)
>>> df['feat_10'] = np.random.randint(2, 5, n)
>>> df['feat_11'] = np.random.randint(1, 4, n)
>>> df['feat_12'] = np.random.randint(0, 4, n)
>>> df['feat_13'] = np.random.normal(6000, 3500, n)
>>> df['feat_14'] = np.random.normal(1000, 500, n)
>>> df['feat_15'] = np.random.normal(1.5, 0.5, n)
>>> df['feat_16'] = np.random.randint(0, 200, n)
>>> df['feat_17'] = np.random.randint(0, 6, n)
>>> df['feat_18'] = np.random.randint(0, 23, n)
>>> df['feat_19'] = pd.Categorical(np.random.choice(['a', 'b', 'c', 'd', 'e'], n), categories=['a', 'b', 'c', 'd', 'e'])
>>> df['feat_20'] = pd.Categorical(np.random.choice(['a', 'b', 'c', 'd', 'e'], n), categories=['a', 'b', 'c', 'd', 'e'])
>>> df['feat_21'] = pd.Categorical(np.random.choice(['a', 'b', 'c', 'd', 'e'], n), categories=['a', 'b', 'c', 'd', 'e'])
>>> df['feat_22'] = pd.Categorical(np.random.choice(['a', 'b', 'c', 'd', 'e'], n), categories=['a', 'b', 'c', 'd', 'e'])
>>> df['feat_23'] = pd.Categorical(np.random.choice([0, 1], n), categories=[0, 1])
>>> df['feat_24'] = pd.Categorical(np.random.choice([0, 1], n), categories=[0, 1])
>>> y = np.random.normal(1500, 500, n)
>>>
>>> print('Y range', min(y), max(y))
Y range -778.1497475199135 3781.263647592668
>>> model = lightgbm.LGBMRegressor(random_state=0, n_estimators=200, objective='mape')
>>> fit = model.fit(df, y, verbose=True)
>>> prediction = fit.predict(df)
>>> print('Y prediction range', min(prediction), max(prediction))
Y prediction range -340.3980111711961 1647.3194057652563
```

I remember there are some bug fixes after 2.2.2 for quantile/mape objectives.

**Comment by MaksimBocharenko** [comment]
Thanks @guolinke for your validation.
I upgraded package and still have the same results.
Also tried in clean virtualenv with python 3.7.2 and python 2.7.15 with only lightgbm and pandas installed.
For all variations I still have the same result.
Tried in docker, same results.

Dockerfile:

FROM python:3.7.2
RUN pip install lightgbm pandas
COPY test_docker.py test_docker.py
CMD python test_docker.py

where test_docker.py includes reproducible example in first comment.
